cpp/ql/src/experimental/semmle/code/cpp/rangeanalysis/extensions/ConstantBitwiseAndExprRange.qll

```diff
@@ -0,0 +1,112 @@
+private import cpp
+private import experimental.semmle.code.cpp.models.interfaces.SimpleRangeAnalysisExpr
+private import semmle.code.cpp.rangeanalysis.RangeAnalysisUtils
+private import experimental.semmle.code.cpp.rangeanalysis.BinaryOrAssignOperation
+
+/**
+ * The current implementation for `BitwiseAndExpr` only handles cases where both operands are
+ * either unsigned or non-negative constants. This class not only covers these cases, but also
+ * adds support for `&` expressions between a signed integer with a non-negative range and a
+ * non-negative constant. It also adds support for `&=` for the same set of cases as `&`.
+ */
+private class ConstantBitwiseAndExprRange extends SimpleRangeAnalysisExpr {
+  BinaryOrAssignConstantBitwiseAndExpr e;
+
+  ConstantBitwiseAndExprRange() { this = e.getOperation() }
```

Comment:
I can see it's convenient to reuse code between `&` and `&=`, but I think it can be done without most of `BinaryOrAssignOperation.qll` and all the one-line wrapper predicates. Here's a sketch of what I have in mind:

```ql
private class ConstantBitwiseAndExprRange extends SimpleRangeAnalysisExpr {
  ConstantBitwiseAndExprRange() {
    (
      this instanceof BitwiseAndExpr
      or
      this instanceof AssignAndExpr
    ) and
   // The stuff with evaluateConstantExpr
  }

  override float getUpperBounds() {
    result =
      getFullyConvertedUpperBounds(getLOp(this))
        .minimum(getFullyConvertedUpperBounds(getROp(this)) and
    // The stuff with hasNegativeRange
  }

  // Plus getLowerBounds and dependsOnChild
}
```

Do you think this can work? Then I think all you need is this class plus the helper predicates `getLOp`, `getROp` and `evaluateConstantExpr`. The `hasNegativeRange` predicate can be inlined if it's simplified to only look at the lower bound.